### PR TITLE
Decrease the uninitialized connection counter as late as possible

### PR DESCRIPTION
**Pull request by 2m**
This fixes (or decreases the chances significantly for) a situation where a stream of incoming connections is closed and the incoming connection does not have a chance to startup properly.

This has been found in [Alpakka test code](https://github.com/akka/alpakka/blob/70f40b1e1f910fadc36a41fdd6e97764a7829ef6/mqtt/src/test/scala/akka/stream/alpakka/mqtt/scaladsl/MqttSourceSpec.scala#L351-L361) where TCP Streams are used to create a proxy server that only handles the first connection.

### CODE REVIEW

**Review by raboof — COMMENTED** [review]
The testcase is very nice and clear :+1: - I find it harder to digest what exactly went wrong and whether this is the way to fix that, will have to dig into that more later

File: akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala [hunk 1]
```diff
@@ -221,11 +218,12 @@ private[stream] object ConnectionSourceStage {
     override def preStart(): Unit = {
       setKeepGoing(true)
       role match {
-        case Inbound(conn, _, _) ⇒
+        case Inbound(conn, _, _, registerCallback) ⇒
           setHandler(bytesOut, readHandler)
           connection = conn
           getStageActor(connected).watch(connection)
           connection ! Register(self, keepOpenOnPeerClosed = true, useResumeWriting = false)
+          registerCallback()
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala]: So before it was decremented already when the stage was materialized but now it's delayed until there is a connection - sounds right to me.

**Review by johanandren — APPROVED** [review]
LGTM, trixy!

File: akka-stream/src/main/mima-filters/2.5.12.backwards.excludes [hunk 2]
```diff
@@ -4,3 +4,10 @@ ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowO
 # #24758 recreate already closed substreams 
 ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.fusing.GroupBy.this")
 ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.stream.scaladsl.FlowOps.groupBy")
+
+# #25135 Decrease the uninitialized connection counter as late as possible
+ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.impl.io.IncomingConnectionStage.this")
```
Comment (raboof) [inline akka-stream/src/main/mima-filters/2.5.12.backwards.excludes]: Do we consider excluding `akka.stream.impl.*` too dangerous? OK in any case, good to double-check exactly what changed I guess.

**Review by raboof — APPROVED** [review]
Seems like an improvement, and outside of the 'hot path'

**Comment by johanandren** [comment]
Fixes #25148

**Pull request merged**